client/app/pages/queries/QuerySource.jsx

```diff
@@ -76,11 +82,25 @@ function QuerySource(props) {
   }), []);
 
   const handleDataSourceChange = useCallback((dataSourceId) => {
-    const newQuery = clone(query);
-    newQuery.data_source_id = dataSourceId;
-    setQuery(newQuery);
+    localStorage.lastSelectedDataSourceId = dataSourceId;
+    if (query.data_source_id !== dataSourceId) {
+      const newQuery = clone(query);
+      newQuery.data_source_id = dataSourceId;
+      setQuery(newQuery);
+    }
   }, [query]);
 
+  useEffect(() => {
+    // choose data source id for new queries
+    if (dataSourcesLoaded && query.isNew()) {
+      const firstDataSourceId = dataSources.length > 0 ? dataSources[0].id : null;
+      handleDataSourceChange(chooseDataSourceId(
+        [query.data_source_id, localStorage.lastSelectedDataSourceId, firstDataSourceId],
```

Comment:
I debugged it in here and there are two things happening that don't allow the localStorage to work: 1 - The localStorage values are always strings; 2 - When the value is `null` it replaces the local storage value, which shouldn't happen IMO as in the first `chooseDataSourceId` executions the DS's are not loaded, thus it will always return `null`.

![image](https://user-images.githubusercontent.com/3356951/70751017-4d309a80-1d0e-11ea-889b-3a1b94480453.png)
-> illustrative image for the 2 above :)

Reply:
I replaced `useEffect` (which is async) with `useMemo` (sync) - this fixed the issue and also reduced amount of component updates. 

https://github.com/getredash/redash/pull/4432/commits/eb53220c1cc818e7cebaea78c8cc695bc575c1ee